src/throttler/throttler.ts

```diff
@@ -254,4 +230,79 @@ export abstract class Throttler<T, R> {
       return p.resolve();
     });
   }
+
+  private initializeTimeout(taskData: TaskData<T, R>, cursorIndex: number): Promise<void> {
+    let rejectFunc: (err: TaskError) => void;
+    const timeoutMillis = taskData.timeoutMillis!;
+    const timeoutPromise = new Promise<void>((_, reject) => {
+      rejectFunc = reject;
+    });
+    taskData.timeoutId = setTimeout(() => {
+      taskData.isTimedOut = true;
+      rejectFunc(new TimeoutError(this.taskName(cursorIndex), timeoutMillis));
+    }, timeoutMillis);
+
+    return timeoutPromise;
+  }
+
+  private async executeTask(taskData: TaskData<T, R>, cursorIndex: number): Promise<any> {
+    const t0 = Date.now();
+    try {
+      const result = await this.handler(taskData.task);
+      const dt = Date.now() - t0;
+      if (dt < this.min) {
+        this.min = dt;
+      }
+      if (dt > this.max) {
+        this.max = dt;
+      }
+      this.avg = (this.avg * this.complete + dt) / (this.complete + 1);
+
+      this.success++;
+      this.complete++;
+      this.active--;
+      return result;
+    } catch (error) {
+      if (taskData.retryCount === this.retries) {
+        throw new RetriesExhaustedError(this.taskName(cursorIndex), this.retries, error);
+      }
+      if (taskData.isTimedOut) {
+        throw error;
+      }
+      this.retried++;
+      taskData.retryCount++;
+      await backoff(taskData.retryCount, this.backoff);
+      logger.debug(`[${this.name}] Retrying task`, this.taskName(cursorIndex));
+      return this.executeTask(taskData, cursorIndex);
+    }
+  }
+
+  private onTaskFulfilled(result: any, taskData: TaskData<T, R>, cursorIndex: number): void {
+    if (taskData.wait) {
+      taskData.wait.resolve(result);
+    }
+    this.cleanupTask(taskData, cursorIndex);
+    this.process();
+  }
+
+  private onTaskFailed(error: TaskError, taskData: TaskData<T, R>, cursorIndex: number): void {
+    this.errored++;
+    this.complete++;
+    this.active--;
+
+    logger.debug(error);
+
+    if (taskData.wait) {
+      taskData.wait.reject(error);
+    }
+    this.cleanupTask(taskData, cursorIndex);
+    this.finish(error);
+  }
+
+  private cleanupTask({ timeoutId }: TaskData<T, R>, cursorIndex: number): void {
+    if (timeoutId) {
+      clearTimeout(timeoutId);
+    }
+    this.taskDataMap.delete(cursorIndex);
```

Comment:
Interestingly, the old implementation only called `taskDataMap.delete(cursorIndex)` when a task had completed successfully, and not when it had failed. But I think I prefer what you've done - calling it in either case.

@mbleigh, do you know if it was deliberate to not delete a task from the `taskDataMap` if the task failed?

Reply:
I...do not remember. My best guess is that it was deliberate because retries? Or is this only called in the event of the final retry failing?

Reply:
My current implementation only deletes the `TaskData` on its final retry failing